Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp

```diff
@@ -156,57 +182,72 @@ static struct wl_buffer* createWaylandBufferFromEGLImage(WPEView* view, WPEBuffe
     }
 
     if (auto* wlBuffer = s_eglCreateWaylandBufferFromImageWL(eglDisplay, eglImage))
-        return wlBuffer;
+        return new DMABufBuffer(wlBuffer);
 
     g_set_error(error, WPE_VIEW_ERROR, WPE_VIEW_ERROR_RENDER_FAILED, "Failed to render buffer: eglCreateWaylandBufferFromImageWL failed with error %#04x", eglGetError());
     return nullptr;
 }
 
 static struct wl_buffer* createWaylandBufferFromDMABuf(WPEView* view, WPEBuffer* buffer, GError** error)
 {
-    if (auto* wlBuffer = static_cast<struct wl_buffer*>(wpe_buffer_get_user_data(buffer)))
-        return wlBuffer;
+    if (auto* dmaBufBuffer = static_cast<DMABufBuffer*>(wpe_buffer_get_user_data(buffer)))
+        return dmaBufBuffer->wlBuffer;
 
-    struct wl_buffer* wlBuffer = nullptr;
+    auto* bufferDMABuf = WPE_BUFFER_DMA_BUF(buffer);
+    DMABufBuffer* dmaBufBuffer = nullptr;
     if (auto* dmabuf = wpeDisplayWaylandGetLinuxDMABuf(WPE_DISPLAY_WAYLAND(wpe_view_get_display(view)))) {
-        auto* bufferDMABuf = WPE_BUFFER_DMA_BUF(buffer);
         auto modifier = wpe_buffer_dma_buf_get_modifier(bufferDMABuf);
         auto* params = zwp_linux_dmabuf_v1_create_params(dmabuf);
         auto planeCount = wpe_buffer_dma_buf_get_n_planes(bufferDMABuf);
         for (guint32 i = 0; i < planeCount; ++i) {
             zwp_linux_buffer_params_v1_add(params, wpe_buffer_dma_buf_get_fd(bufferDMABuf, i), i, wpe_buffer_dma_buf_get_offset(bufferDMABuf, i),
                 wpe_buffer_dma_buf_get_stride(bufferDMABuf, i), modifier >> 32, modifier & 0xffffffff);
         }
-        wlBuffer = zwp_linux_buffer_params_v1_create_immed(params, wpe_buffer_get_width(buffer), wpe_buffer_get_height(buffer),
+        auto* wlBuffer = zwp_linux_buffer_params_v1_create_immed(params, wpe_buffer_get_width(buffer), wpe_buffer_get_height(buffer),
             wpe_buffer_dma_buf_get_format(bufferDMABuf), 0);
         zwp_linux_buffer_params_v1_destroy(params);
 
         if (!wlBuffer) {
             g_set_error_literal(error, WPE_VIEW_ERROR, WPE_VIEW_ERROR_RENDER_FAILED, "Failed to render buffer: failed to create Wayland buffer from DMABuf");
             return nullptr;
         }
+
+        dmaBufBuffer = new DMABufBuffer(wlBuffer);
     } else {
-        wlBuffer = createWaylandBufferFromEGLImage(view, buffer, error);
-        if (!wlBuffer)
+        dmaBufBuffer = createWaylandBufferFromEGLImage(view, buffer, error);
+        if (!dmaBufBuffer)
             return nullptr;
     }
 
-    wl_buffer_add_listener(wlBuffer, &bufferListener, buffer);
+    auto* toplevel = wpe_view_get_toplevel(WPE_VIEW(view));
+    if ((toplevel && !wpeToplevelWaylandHasSyncObject(WPE_TOPLEVEL_WAYLAND(toplevel))) || wpe_buffer_dma_buf_get_rendering_fence(bufferDMABuf) == -1)
```

Comment:
render buffer can't be called without a toplevel, so we don't need to null-check it here.